recipe/session/recipeImplementation.go

```diff
@@ -118,28 +220,89 @@ func makeRecipeImplementation(querier supertokens.Querier, config sessmodels.Typ
+		for _, tokenTransferMethod := range availableTokenTransferMethods {
+			token, err := getToken(req, sessmodels.RefreshToken, tokenTransferMethod)
+			if err != nil {
+				return nil, err
+			}
+			refreshTokens[tokenTransferMethod] = token
+			if token != nil {
+				supertokens.LogDebugMessage("refreshSession: got refresh token from " + string(tokenTransferMethod))
+			}
 		}
 
-		inputRefreshToken := getRefreshTokenFromCookie(req)
-		if inputRefreshToken == nil {
-			supertokens.LogDebugMessage("refreshSession: UNAUTHORISED because refresh token from cookies is undefined")
-			return nil, errors.UnauthorizedError{Msg: "Refresh token not found. Are you sending the refresh token in the request as a cookie?"}
+		allowedTokenTransferMethod := config.GetTokenTransferMethod(req, false, userContext)
+		supertokens.LogDebugMessage("refreshSession: getTokenTransferMethod returned " + string(allowedTokenTransferMethod))
+
+		var requestTokenTransferMethod sessmodels.TokenTransferMethod
+		var refreshToken *string
+
+		if (allowedTokenTransferMethod == sessmodels.Any || allowedTokenTransferMethod == sessmodels.Header) && refreshTokens[sessmodels.Header] != nil {
+			supertokens.LogDebugMessage("refreshSession: using header transfer method")
+			requestTokenTransferMethod = sessmodels.Header
+			refreshToken = refreshTokens[sessmodels.Header]
+		} else if (allowedTokenTransferMethod == sessmodels.Any || allowedTokenTransferMethod == sessmodels.Cookie) && refreshTokens[sessmodels.Cookie] != nil {
+			supertokens.LogDebugMessage("refreshSession: using cookie transfer method")
+			requestTokenTransferMethod = sessmodels.Cookie
+			refreshToken = refreshTokens[sessmodels.Cookie]
+		} else {
+			if getCookieValue(req, LEGACY_ID_REFRESH_TOKEN_COOKIE_NAME) != nil {
+				supertokens.LogDebugMessage("refreshSession: cleared legacy id refresh token because refresh token was not found")
+				setCookie(config, res, LEGACY_ID_REFRESH_TOKEN_COOKIE_NAME, "", 0, "accessTokenPath")
+			}
+
+			supertokens.LogDebugMessage("refreshSession: UNAUTHORISED because refresh token in request is undefined")
+			False := false
+			return nil, errors.UnauthorizedError{
+				Msg:         "Refresh token not found. Are you sending the refresh token in the request as a cookie?",
+				ClearTokens: &False,
+			}
 		}
 
 		antiCsrfToken := getAntiCsrfTokenFromHeaders(req)
-		response, err := refreshSessionHelper(recipeImplHandshakeInfo, config, querier, *inputRefreshToken, antiCsrfToken, getRidFromHeader(req) != nil)
+		response, err := refreshSessionHelper(recipeImplHandshakeInfo, config, querier, *refreshToken, antiCsrfToken, getRidFromHeader(req) != nil, requestTokenTransferMethod)
 		if err != nil {
+			// This token isn't handled by getToken/setToken to limit the scope of this legacy/migration code
+			if !defaultErrors.As(err, &errors.UnauthorizedError{}) {
+				if getCookieValue(req, LEGACY_ID_REFRESH_TOKEN_COOKIE_NAME) != nil {
```

Comment:
I think it may be more concise if you extracted the `getCookieValue(req, LEGACY_ID_REFRESH_TOKEN_COOKIE_NAME)` check into a variable.

Reply:
Simplified the condition and removed duplicates.